### DISCUSSION TITLE
[General] #Building DEX Trade Analysis Tool on zkSync Era - Need Help with Transaction Type Detection

### DISCUSSION THREAD

**Discussion by 0xNexius** [zkSync-Community-Hub/zksync-developers#786] [General] [answered]
### Environment

Mainnet

## Project Goal
Building a tool to analyze wallet addresses and identify "paperhanded" trades. We need to:
- Track all token transactions for a given wallet
- Identify sell transactions and their prices
- Compare sell prices with ATH price
- Calculate potential gains missed by selling early

### 1. Transaction Classification Issues
Currently struggling to differentiate between:
- Actual token sells on DEXes
- Regular token transfers between wallets
- Liquidity providing/removing operations
- Staking/unstaking transactions

Problems:
- Can't reliably identify when a wallet sells tokens
- DEX router contract calls failing during verification
- Missing sells through newer or less common DEXes
- False positives: transfers/LP/staking being marked as sells

### 2. Historical Price Tracking
Need to calculate sell prices and subsequent price movements.

Problems:
- Can't accurately determine token price at time of sell
- No reliable way to track price movements after sells
- Missing data for calculating potential gains lost
- Price calculation fails when transaction involves multiple tokens

### 3. DEX Coverage Issues
Need to identify all possible ways a user might sell tokens:
- Standard DEX swaps
- LP removals that could be considered "selling"
- Complex transactions involving multiple DEXes
- New DEXes being launched regularly

## Key Questions

1. **Transaction Type Detection**
   - What's the most reliable way to differentiate between sells and other operations (transfers/LP/staking)?
   - How can we handle complex transactions that might involve multiple operations?
   - What specific contract interactions indicate a true "sell"?

2. **Price Tracking**
   - How can we accurately determine the price at which tokens were sold?
   - Is there a way to track price movements after a sell?
   - How can we handle price calculation for complex swap paths?

3. **DEX Coverage**
   - What are all the major DEXes we should monitor for sells?
   - How can we handle new DEXes as they launch?
   - How to differentiate DEX trades from other protocol interactions?

The goal is to accurately identify when users paperhanded (sold too early) and calculate how much they would have made by selling at ATH. Any guidance on these challenges would be greatly appreciated!

**Comment by alhenaai** [comment]
For building a DEX trade analysis tool on ZKsync Era that can accurately classify transaction types and track historical prices, here are some suggestions based on the challenges you've outlined:

### 1. **Transaction Type Detection**
- **Smart Contract Analysis:** Utilize tools like [Elliptic](https://www.elliptic.co/) for real-time wallet and transaction screening. This can help in distinguishing between different types of transactions by analyzing the data patterns and contract interactions.
- **Heuristic Methods:** Develop heuristics based on common patterns in DEX transactions, such as identifying the DEX router addresses and the method signatures used for swaps, liquidity additions, and removals.
- **Event Logs:** Analyze the event logs emitted by smart contracts during transactions. Sell transactions typically emit specific events that can be distinguished from other types of interactions.

### 2. **Historical Price Tracking**
- **Use Oracle Services:** Integrate price oracles that provide historical data, such as Chainlink, to fetch the price at the time of the transaction.
- **DEX APIs:** Leverage APIs from major DEX platforms that provide historical trade data, which can be useful for reconstructing the price path of a token.

### 3. **DEX Coverage**
- **Comprehensive List Monitoring:** Regularly update your list of DEXes based on new launches and popular platforms. Monitoring platforms like [Coingecko](https://www.coingecko.com/en/categories/zksync-ecosystem) can be useful to keep track of active DEXes within the ZKsync ecosystem.
- **Dynamic Interaction Tracking:** Develop a system to dynamically identify and classify interactions with new DEX contracts by monitoring for new contract creations and transactions that resemble DEX interactions.

For more detailed discussions and community support, you can visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions). This platform can provide further insights and peer advice on handling the complexities of your project.

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by uF4No** [comment, marked as answer]
Hi

Sharing some info although this page is more focused on technical support and debugging issues with our tools. Perhaps you can  also [ask in our Discord server](https://join.zksync.dev/) as well to get better answers.

**Transaction Type Detection**

> What's the most reliable way to differentiate between sells and other operations (transfers/LP/staking)?

As far as I know, a sell will be just a transfer from the user account to another one. I'd try to identify any transactions on the token contract, calling the `transfer` method with the user address in the `from` param. [See the Openzeppelin docs for the ERC20 token API](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-transfer-address-uint256-).

> What specific contract interactions indicate a true "sell"?

As mentioned before, any `transfer` where the `from` is the user's account.

**Price Tracking**

> How can we accurately determine the price at which tokens were sold?

You should be able to see the number of tokens going out of the user's wallet and the amount of ETH or other tokens going in. And you can use different oracles to get the USD price of different tokens.

> Is there a way to track price movements after a sell?

You can use APIs like Coingecko or Alchemy

**DEX coverage**

> What are all the major DEXes we should monitor for sells?

On ZKsync Era currently Syncswap, Koi, zkSwap, Maverick and Uniswap. See [DefiLlama for more details](https://defillama.com/chain/ZKsync%20Era)

> How can we handle new DEXes as they launch?

Monitor DefiLlama and [the official ZKsync X account](https://x.com/zksync)
